runtime/services/catalog/migrator/sources/sources.go

```diff
@@ -176,50 +176,168 @@ func (m *sourceMigrator) ExistsInOlap(ctx context.Context, olap drivers.OLAPStor
+		t, ok = srcConnector.AsTransporter(srcConnector, olapConnection)
+		if !ok {
+			return fmt.Errorf("data transfer not possible from %q to %q", srcConnector.Driver(), olapConnection.Driver())
+		}
 	}
 
-	ingestionSummary, err := olap.Ingest(ctx, env, source)
+	src, err := source(apiSource.Connector, apiSource)
 	if err != nil {
 		return err
 	}
 
-	catalogObj.BytesIngested = ingestionSummary.BytesIngested
-	return nil
+	sink := sink(olapConnection.Driver(), name)
+
+	timeout := _defaultIngestTimeout
+	if apiSource.GetTimeoutSeconds() > 0 {
+		timeout = time.Duration(apiSource.GetTimeoutSeconds()) * time.Second
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	ingestionLimit := opts.IngestStorageLimitInBytes
+	p := &progress{}
+	done := make(chan bool, 1)
+	ticker := time.NewTicker(5 * time.Second)
+	limitExceeded := false
+	go func() {
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
+			olap, _ := olapConnection.AsOLAPStore()
+			if size, ok := olap.EstimateSize(); ok && size > ingestionLimit {
+				limitExceeded = true
+				cancel()
+			}
+		}
+	}()
+	err = t.Transfer(ctxWithTimeout, src, sink, drivers.NewTransferOpts(drivers.WithLimitInBytes(ingestionLimit)), p)
+	done <- true
+	if limitExceeded {
+		return drivers.ErrIngestionLimitExceeded
+	}
+	return err
+}
+
+type progress struct {
+	catalogObj drivers.CatalogEntry
+	unit       drivers.ProgressUnit
+}
+
+func (p *progress) Target(val int64, unit drivers.ProgressUnit) {
+	p.unit = unit
+}
+
+func (p *progress) Observe(val int64, unit drivers.ProgressUnit) {
+	if unit == drivers.ProgressUnitByte {
+		p.catalogObj.BytesIngested += val
+	}
+}
+
+func source(connector string, src *runtimev1.Source) (drivers.Source, error) {
+	props := src.Properties.AsMap()
+	switch connector {
+	case "s3":
+		return &drivers.BucketSource{
+			ExtractPolicy: src.Policy,
+			Properties:    props,
+		}, nil
+	case "gcs":
+		return &drivers.BucketSource{
+			ExtractPolicy: src.Policy,
+			Properties:    props,
+		}, nil
+	case "https":
+		return &drivers.FileSource{
+			Properties: props,
+		}, nil
+	case "local_file":
+		return &drivers.FileSource{
+			Properties: props,
+		}, nil
+	case "motherduck":
+		query, ok := props["query"]
+		if !ok {
+			return nil, fmt.Errorf("property \"query\" is mandatory for connector \"motherduck\"")
+		}
+		var db string
+		if val, ok := props["db"]; ok {
+			db = val.(string)
+		}
+
+		return &drivers.DatabaseSource{
+			Query:    query.(string),
+			Database: db,
+		}, nil
+	default:
+		return nil, nil
+	}
+}
+
+func sink(connector, tableName string) drivers.Sink {
+	switch connector {
+	case "duckdb":
+		return &drivers.DatabaseSink{
+			Table: tableName,
+		}
+	default:
+		return nil
+	}
+}
+
+// TODO :: may be pass all envs ?
+// It is slightly confusing right now that what gets passed to driver.Open vs source property
+func connectorVariables(src *runtimev1.Source, env map[string]string, repoRoot string) map[string]any {
```

Comment:
Agreed on the confusion. Hopefully some of the other refactors for connector config will help alleviate it